File: go/verifier.go
package partnersigned

import (
	"bytes"
	"context"
	"errors"
	"io"
	"net/http"
	"strconv"
	"time"
)

// DefaultTolerance is the allowed clock skew between sender and receiver (±300s).
const DefaultTolerance = 300 * time.Second

// DefaultMaxBodyBytes caps how much of a request body Verifier.VerifyRequest reads.
const DefaultMaxBodyBytes int64 = 1 << 20

// Errors returned by Verifier.
var (
	ErrMissingHeaders   = errors.New("partnersigned: missing X-Timestamp, X-Request-Id or X-Signature")
	ErrInvalidTimestamp = errors.New("partnersigned: X-Timestamp is not a Unix timestamp")
	ErrStaleTimestamp   = errors.New("partnersigned: X-Timestamp is outside the allowed window")
	ErrInvalidSignature = errors.New("partnersigned: signature does not match")
)

// Verifier is the receiving-side check for a signed request. Use it in your own /deduct endpoint
// to validate what BanglaReels sends you. It mirrors the backend's PartnerSignatureGuard: first the
// timestamp window, then the signature over the exact raw body.
type Verifier struct {
	// Scheme is SchemeHMAC or SchemeEd25519.
	Scheme Scheme
	// Key is the HMAC secret, or the sender's Ed25519 PUBLIC key PEM.
	Key string
	// Tolerance is the allowed clock skew. Zero means DefaultTolerance.
	Tolerance time.Duration
	// Now returns the current time. Nil means time.Now.
	Now func() time.Time
	// MaxBodyBytes caps the body VerifyRequest reads. Zero means DefaultMaxBodyBytes.
	MaxBodyBytes int64
}

// Verify checks the three signature headers against rawBody, which must be the request body
// exactly as received. It returns nil when the request is valid.
func (v *Verifier) Verify(timestamp, requestID, signature string, rawBody []byte) error {
	if timestamp == "" || requestID == "" || signature == "" {
		return ErrMissingHeaders
	}
	sentAt, err := strconv.ParseInt(timestamp, 10, 64)
	if err != nil {
		return ErrInvalidTimestamp
	}

	now := time.Now
	if v.Now != nil {
		now = v.Now
	}
	tolerance := v.Tolerance
	if tolerance == 0 {
		tolerance = DefaultTolerance
	}
	skew := now().Unix() - sentAt
	if skew < 0 {
		skew = -skew
	}
	if skew > int64(tolerance/time.Second) {
		return ErrStaleTimestamp
	}

	if !Verify(v.Scheme, BuildSigningString(timestamp, requestID, rawBody), signature, v.Key) {
		return ErrInvalidSignature
	}
	return nil
}

// VerifyRequest reads r.Body, verifies it together with the X-Timestamp, X-Request-Id and
// X-Signature headers, and returns the raw body so you can decode it afterwards. r.Body is
// replaced with a fresh reader over the same bytes.
func (v *Verifier) VerifyRequest(r *http.Request) ([]byte, error) {
	limit := v.MaxBodyBytes
	if limit == 0 {
		limit = DefaultMaxBodyBytes
	}

	var rawBody []byte
	if r.Body != nil {
		var err error
		rawBody, err = io.ReadAll(io.LimitReader(r.Body, limit))
		r.Body.Close()
		if err != nil {
			return nil, err
		}
	}
	r.Body = io.NopCloser(bytes.NewReader(rawBody))

	err := v.Verify(
		r.Header.Get(HeaderTimestamp),
		r.Header.Get(HeaderRequestID),
		r.Header.Get(HeaderSignature),
		rawBody,
	)
	if err != nil {
		return nil, err
	}
	return rawBody, nil
}

type rawBodyKey struct{}

// Middleware rejects requests that fail VerifyRequest with 401 Unauthorized. For requests that
// pass, the raw body is available both as r.Body and via RawBodyFromContext.
func (v *Verifier) Middleware(next http.Handler) http.Handler {
	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		rawBody, err := v.VerifyRequest(r)
		if err != nil {
			http.Error(w, "invalid signature", http.StatusUnauthorized)
			return
		}
		ctx := context.WithValue(r.Context(), rawBodyKey{}, rawBody)
		next.ServeHTTP(w, r.WithContext(ctx))
	})
}

// RawBodyFromContext returns the verified raw body stored by Verifier.Middleware.
func RawBodyFromContext(ctx context.Context) ([]byte, bool) {
	b, ok := ctx.Value(rawBodyKey{}).([]byte)
	return b, ok
}

File: rust/src/verifier.rs
//! Receiving-side check for a signed request — use it in your own `/deduct` endpoint to validate
//! what BanglaReels sends you. Mirrors what the backend's PartnerSignatureGuard does for the
//! into-BanglaReels direction: timestamp window, then signature over the exact raw body.

use std::time::{Duration, SystemTime, UNIX_EPOCH};

use reqwest::header::HeaderMap;

use crate::crypto::{self, SigningScheme};
use crate::error::Result;

/// Allowed clock skew between sender and receiver (±300s).
pub const DEFAULT_TOLERANCE: Duration = Duration::from_secs(300);

/// The three signature headers of an incoming request.
#[derive(Debug, Clone, Copy, Default, PartialEq, Eq)]
pub struct SignedHeaders<'a> {
    /// `X-Timestamp` (Unix seconds).
    pub timestamp: Option<&'a str>,
    /// `X-Request-Id`.
    pub request_id: Option<&'a str>,
    /// `X-Signature` (`v1=...`).
    pub signature: Option<&'a str>,
}

impl<'a> SignedHeaders<'a> {
    pub fn new(timestamp: &'a str, request_id: &'a str, signature: &'a str) -> Self {
        Self {
            timestamp: Some(timestamp),
            request_id: Some(request_id),
            signature: Some(signature),
        }
    }

    /// Reads `X-Timestamp`, `X-Request-Id` and `X-Signature` from an `http` header map (the type
    /// used by reqwest, axum, hyper, ...). Missing or non-UTF-8 headers become `None`.
    pub fn from_header_map(headers: &'a HeaderMap) -> Self {
        let get = |name: &str| headers.get(name).and_then(|v| v.to_str().ok());
        Self {
            timestamp: get("x-timestamp"),
            request_id: get("x-request-id"),
            signature: get("x-signature"),
        }
    }
}

/// Verifies an incoming signed request against the current time with the default ±300s window.
///
/// - `key`: HMAC secret, or the sender's Ed25519 PUBLIC key PEM.
/// - `raw_body`: the request body EXACTLY as received — read it before deserializing.
///
/// Returns `Ok(false)` for missing headers, a stale/unparseable timestamp or a bad signature.
pub fn verify_request(
    scheme: SigningScheme,
    key: &str,
    headers: &SignedHeaders<'_>,
    raw_body: impl AsRef<[u8]>,
) -> Result<bool> {
    let now = SystemTime::now()
        .duration_since(UNIX_EPOCH)
        .map(|d| d.as_secs() as i64)
        .unwrap_or_default();
    verify_request_at(scheme, key, headers, raw_body, now, DEFAULT_TOLERANCE)
}

/// Like [`verify_request`], with an explicit "now" (Unix seconds) and tolerance — handy for tests.
pub fn verify_request_at(
    scheme: SigningScheme,
    key: &str,
    headers: &SignedHeaders<'_>,
    raw_body: impl AsRef<[u8]>,
    now_unix_secs: i64,
    tolerance: Duration,
) -> Result<bool> {
    let (Some(timestamp), Some(request_id), Some(signature)) =
        (headers.timestamp, headers.request_id, headers.signature)
    else {
        return Ok(false);
    };
    if timestamp.is_empty() || request_id.is_empty() || signature.is_empty() {
        return Ok(false);
    }

    let Ok(sent_at) = timestamp.parse::<i64>() else {
        return Ok(false);
    };
    if now_unix_secs.abs_diff(sent_at) > tolerance.as_secs() {
        return Ok(false);
    }

    let signing_string = crypto::build_signing_string(timestamp, request_id, raw_body);
    crypto::verify(scheme, &signing_string, signature, key)
}
